### PR TITLE
connection: try all addresses on connecting

### PR DESCRIPTION
**Pull request by tisonkun**
This closes #134.

Signed-off-by: tison <wander4096@gmail.com>
Co-authored-by: Marc-Antoine Perennou <Marc-Antoine@Perennou.com>

### CODE REVIEW

File: rust-toolchain.toml
Comment (tisonkun) [inline rust-toolchain.toml]: I don't find documents about MSRV so just assume that we use the nightly version as usual.

Add this file as it's required to enable feature gates (`drain_filter` in this patch). Although it's possible to write a utility function, if we can just use stdlib, I'd prefer to use it.

**Comment by tisonkun** [comment]
ping @FlorentinDUBOIS

File: rust-toolchain.toml
Comment (FlorentinDUBOIS) [reply rust-toolchain.toml]: To me, the toolchain have to be stable, if it is not set. I am strongly discourage nightly requirement on production driver.

File: src/connection.rs
Comment (FlorentinDUBOIS) [inline src/connection.rs]: To me, this line should be written like below to remove the nightly requirement:
```suggestion
            .drain()
            .filter(|e| e.establish_retryable())
```

**Review by FlorentinDUBOIS — CHANGES_REQUESTED** [review]
Thank for the pull request, but I cannot merge since it requires nightly rust build.

**Comment by tisonkun** [comment]
@FlorentinDUBOIS OK. I will point the toolchain to stable and find an alternative to the related code.

**Comment by FlorentinDUBOIS** [comment]
Thank you! I provided you a way to do it, I hope it will help you :)

**Comment by FlorentinDUBOIS** [comment]
Once, your modifications will be merge, I am quite motivated to release a 4.2.0

File: src/connection.rs
Comment (tisonkun) [reply src/connection.rs]: Not exactly. `drain` take a range instead of a condition. And if we move the original errors, we cannot use it in the line:

```rust
error!("connection error, not retryable: {:?}", errors);
```

But perhaps we don't have to do error handling so fine-grained.

Comment (tisonkun) [reply src/connection.rs]: But we can simply write some imperative code here. I'm preparing a commit on this.

**Comment by tisonkun** [comment]
@FlorentinDUBOIS fixup pushed and tested locally. Let's wait for the CI verification ;)

File: rust-toolchain.toml
Comment (tisonkun) [inline rust-toolchain.toml:2]: @FlorentinDUBOIS Now it means the latest stable version. I think it's OK for now and we may choose an fixed MSRV when #224 is finished and we think we've completed the major feature set.

**Review by FlorentinDUBOIS — APPROVED** [review]

File: rust-toolchain.toml
Comment (FlorentinDUBOIS) [reply rust-toolchain.toml:2]: I am ok with that, but nightly is really tricky :sweat_smile:

**Pull request merged**
